fix($interpolate): use custom toString() function & fix(ngBind): use same json conversion as $interpolate

Narretz opened the pull request:
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
bug fix

**What is the current behavior? (You can also link to an open issue here)**
ngBind has different binding logic for objects than interpolation

**What is the new behavior (if this is a feature change)?**
Same logic as interpolate

**Does this PR introduce a breaking change?**
Yes, the visible output is now different, i.e. consistent with interpolate. I don't think anyone relied on ngBind producing this output.
However, this PR is a step away from the proposal in #11406, which wants $interpolate to call toString() instead of toJson. See also PR #8350, and issue #7317

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been updated (we should probably document how this works in the guide)

**Other information**:

Fixes #11716

BREAKING CHANGE:

`ngBind` now uses the same logic as $interpolate (i.e. {{myString}}) when
binding, which means objects are now transformed with JSON.stringify. Before, it
would use the object's toString() function.

The following example shows the different output:

``` js
$scope.myObject = {a: 1, b: 2};
```

``` html
<span ng-bind="myObject"></span>
```

Before:

``` html
<span ng-bind="myObject">[object Object]</span>
```

After:

``` html
<span ng-bind="myObject">{"a":1,"b":2}</span>
```

Narretz committed:
fix(ngBind): use same string representation as $interpolate

Fixes #11716

BREAKING CHANGE:

`ngBind` now uses the same logic as $interpolate (i.e. {{myString}}) when
binding, which means values other than strings are now transformed as following:
- null / undefined become empty string
- with an object's custom toString() function, except if the object is a Date, Array, or Number
- otherwise with JSON.stringify

Previously, ngBind would use always use toString().

The following examples show the different output:
```js
$scope.myPlainObject = {a: 1, b: 2};
$scope.myCustomObject = {a: 1, b: 2, toString: function() {return 'a+b';}};
```

Plain Object:
```html
<!-- Before: -->
<span ng-bind="myPlainObject">[object Object]</span>

<!-- After: -->
<span ng-bind="myPlainObject">{"a":1,"b":2}</span>
```

Object with custom toString():

```html
<!-- Before: -->
<span ng-bind="myCustomObject">[object Object]</span>

<!-- After: -->
<span ng-bind="myCustomObject">a+b</span>
```

If you want the output of `toString()`, you can use it directly on the value in ngBind:

```html
<span ng-bind="myObject.toString()">[object Object]</span>
```

campersau commented on src/ng/directive/ngBind.js:
```diff
@@ -60,7 +60,7 @@ var ngBindDirective = ['$compile', function($compile) {
         element = element[0];
         scope.$watch(attr.ngBind, function ngBindWatchAction(value) {
-          element.textContent = isUndefined(value) ? '' : value;
+          element.textContent = isUndefined(value) ? '' : stringify(value);
```
This can be simplified to:
```js
element.textContent = stringify(value);
```

petebacondarwin replied:
Can you include this change too @Narretz ?

gkalpak commented:
I think it makes sense to make `ngBind` and `$interpolate` consistent in terms of object stringification, but moving from `toString` to `toJson` will be a step backwards for oobjects with custom `toString` methods.

I think a good compromise would be to use `toString` if it different than `Object.prototype.toString` and `toJson` otherwise (either only on `ngBind` or on `$interpolate` too - but that's a bigger breaking change).

petebacondarwin commented:
I agree with @gkalpak that we should use `toString` if a custom one is available. Given that this is a BC anyway. Let's change `$interpolate` to do this too and only merge to 1.6

Narretz commented:
I've updated the PR to include the commit that calls custom toString() on objects. PTAL.

I've noticed a problem with our own date input tests. In the original PR, Date objects als had toString() called when interpolating them. This change made our tests fail that allow min and max attributes of date inputs to be date objects: https://github.com/angular/angular.js/blob/f3377da6a748007c11fde090890ee58fae4cefa5/test/ng/directive/inputSpec.js#L1896. We expect a jsonified string like: `'"2014-11-09T23:00:00.000Z"'`. Date.toString() however produces a locale aware string that looks like this:  `Mon Nov 10 2014 00:00:00 GMT+0100 (Mitteleuropäische Zeit), `

As a fix, I've updated the createDateParser function to convert the passed string into an object if manual iso date parsing fails. This will produce a valid Date from the result of Date.toString(), and NaN from everything that can't be parsed (NaN is also returned if nothing works).

There's also the issue that ngMessageFormat duplicates the stringify function, and at the moment all our unit tests will use this function instead of the one defined in Angular.js. This is a bit weird, but at least it guarantees that you notice when you have made changes to the core implementation.

gkalpak commented on src/Angular.js:
```diff
@@ -1929,6 +1930,28 @@ function createMap() {
+      value = '' + value;
+      break;
+    default:
+      if (isFunction(value.toString) && value.toString !== Object.prototype.toString && !isArray(value)) {
```
I think we already have `toString` as an "IIFE global".

Narretz replied:
We do have a hasCustomToString fn

gkalpak replied:
Even better :smiley:

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -1290,6 +1290,9 @@ function createDateParser(regexp, mapping) {
         return new Date(map.yyyy, map.MM - 1, map.dd, map.HH, map.mm, map.ss || 0, map.sss * 1000 || 0);
+      } else {
+        // If we fail to parse the string with the ISO regex, it might be because it's the result of Date.toString()
+        return Date.parse(iso);
```
This can lead to unexpected results. E.g. `Date.parse('0')` is the same as `Date.parse('2000')` which is the same as `2000-01-01T00:00:00` in local time - which is probably not the intended value :smiley:

If this is only to facilitate our new `toString()` stringification?

Narretz replied:
Yes, this is only the make it work with Date.toString(). Maybe we can only use Date.parse() when the string has a minimum length of x?

gkalpak replied:
If I understand correctly, this change was made in order to fix the tests. I don't think it's the right fix.

We wanted to test that JSONified dates would be parsed correctly. In order to test that, we relied on the fact that `$interpolate` would convert date objects to JSON. Now that `$interpolate` does not convert them to JSON (but calls `toString()` instead), we should change the _tests_, so that JSONified dates are still passed to `max`/`min`.

With the current fix, we have a test that is supposed to test parsing JSONified dates, but it tests stringified dates instead, right ?

--
Part of the big BC will be that `max`/`min` won't be able to handle interpolated date objects. One will need to either use `ngMax/Min` or change `{{ myDateObj }}` to `{{ myDateObj.toISOString() }}` or `{{ myDateObj | json }}` (or manually JSONify the date object in the controller).
I don't think it's too bad, though.

Another idea would be to check if `new Date(iso).toString() === iso`. If this is true, then `iso` must be the `toString()` output of a Date equal to `new Date(iso)`.

--
BTW, `Date.parse()` returns the number of milliseconds. Don't we need a Date object to be returned ?

Narretz replied:
I'm not 100% sure, but maybe you are on to something here.
The tests say "should parse ISO-based date strings as a valid min date value", but they specifically use Date objects. The commit that introduced that - https://github.com/angular/angular.js/commit/088545c1856ce1c3ec3416965dff65077a6e0523 - also said it closes #6755, but it looks to me that it should say Closes #6754, which is precisely about using Min/Max with Date objects: https://github.com/angular/angular.js/issues/6754
For #6755, there's another PR: https://github.com/angular/angular.js/commit/b3502835039178296b730b7526e5666b66ba9156

So I assume the intention was to allow JSONified Date objects; however this change was never documented.
I'd say that we should make it work and document it. I like your comparson approach: `new Date(iso).toString() === iso`

I assumed `Date.parse()` returns a Date object -_-. Yes, we probably want new Date().

petebacondarwin replied:
I think that the 6754 and 6755 closes were just the wrong way round.

Changing the failing tests to use `toISOString()` prevents the hack in `input.js`.

The question is whether we need additional tests that specifically check that we can use real `Date` objects as well?

petebacondarwin replied:
The following test passes, without the hack:

```
    it('should validate ng-min values that are Date objects correctly', function() {
      $rootScope.min = new Date(2013, 1, 1, 1, 2, 0);
      var inputElm = helper.compileInput('<input type="datetime-local" ng-model="value" name="alias" ng-min="min" />');

      helper.changeInputValueTo('2010-01-01T12:34:00');
      expect(inputElm).toBeInvalid();

      $rootScope.min = new Date(2014, 1, 1, 1, 2, 0);
      $rootScope.$digest();

      expect(inputElm).toBeInvalid();

      $rootScope.min = new Date(2009, 1, 1, 1, 2, 0);
      $rootScope.$digest();

      expect(inputElm).toBeValid();
    });
```

Narretz replied:
#6754 was specifically about `min` and `max` using date objects, not the `ng-` variants. Because you are missing out on native validation if you use the `ng` variants. So I don't think the test ensures that we support what the commit claimed to support.

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -1358,17 +1361,21 @@ function createDateInputType(type, regexp, parseDate, format) {
 
+function isValidDate(value) {
+  // Invalid Date: getTime() returns NaN
+  return value && !(value.getTime && value.getTime() !== value.getTime());
```
Unrelated to this PR: 
That's a weird validation function (e.g. `isValidDate({})` -- true).
I would expect `value && value.getTime && value.getTime() === value.getTime()`.

gkalpak commented on test/ng/directive/ngBindSpec.js:
```diff
@@ -46,6 +46,15 @@ describe('ngBind*', function() {
+        expect(element.text()).toEqual(prop[1]);
+      });
+    });
+
```
I would add a test with a non-default `toString` here too.

gkalpak commented on test/ng/interpolateSpec.js:
```diff
@@ -35,6 +35,22 @@ describe('$interpolate', function() {
   }));
 
+  it('should use custom toString when present', inject(function($interpolate, $rootScope) {
+     var context = {
```
Wrong indentation :stuck_out_tongue:

petebacondarwin commented:
Perhaps we should move `stringify` into its own file, which is included in both `angular.js` and `angular-message-format.js`? That way we have only one version in the source code (and for testing).

petebacondarwin commented:
Actually there already IS a `stringify.js` file that is shared with the `angular-loader.js` file. So we should make use of that, right?

Narretz commented:
Yes, but we need to share between ng and ngMessageFormat and it looks like ngMessageFormat only needs the stringify and not the other functions

Narretz committed:
test(input): ensure Date objects work for min/max in date input types

Tests that
- interpolated Date objects work for min/max
- Date objects work for ng-min/ng-max

Narretz committed:
feat($interpolate): use custom toString() function if present

Except on Numbers, Dates and Arrays.

Thanks to @danielkrainas for the initial implementation of this feature.

This behavior is consistent with implementations found in other languages such as Ruby, Python,
and CoffeeScript.
http://rubymonk.com/learning/books/1-ruby-primer/chapters/5-strings/lessons/31-string-basics
https://docs.python.org/2/library/stdtypes.html#string-formatting-operations
http://coffeescriptcookbook.com/chapters/strings/interpolation

The commit also exposes a private $$stringify method on the angular global, so that ngMessageFormat
can use the same logic without duplicating it.

Fixes #7317
Closes #8350
Fixes #11406

BREAKING CHANGE:

When converting values to strings, interpolation now uses a custom toString() function on objects
that are not Number, Array or Date (custom means that the `toString` function is not the same as
`Object.prototype.toString`). Otherwise, interpolation uses JSON.stringify() as usual.

Should you have a custom toString() function but still want the output of JSON.stringify(),
migrate as shown in the following examples:

Before:

```html
<span>{{myObject}}</span>
```

After - use the `json` filter to stringify the object:

```html
<span>{{myObject | json}}</span>
```

petebacondarwin commented:
One nitpic but otherwise LGTM

Narretz commented:
What's that nitpick @petebacondarwin ?

Narretz commented:
PR has been updated, tests pass, PTAL

petebacondarwin commented on test/ng/directive/ngBindSpec.js:
```diff
@@ -78,7 +78,7 @@ describe('ngBind*', function() {
       element = $compile('<div ng-bind="bind"></div>')($rootScope);
       $rootScope.$digest();
-      expect(element.text()).toEqual('"2014-11-09T23:00:00.000Z"');
+      expect(element.text()).toBe('"' + $rootScope.bind.toJSON() + '"');
```
Unrelated to this change but, is it me, or does it seem weird to use a property called `bind` in this test?

petebacondarwin commented on test/ng/interpolateSpec.js:
```diff
@@ -53,7 +53,8 @@ describe('$interpolate', function() {
   it('should NOT use toString on Date objects', inject(function($interpolate) {
-    expect($interpolate('{{a}}')({ a: new Date(2014, 10, 10, 0, 0, 0) })).toEqual('"2014-11-09T23:00:00.000Z"');
+    var date = new Date(2014, 10, 10);
+    expect($interpolate('{{a}}')({ a: date })).toBe('"' + date.toJSON() + '"');
```
Perhaps, keeping more inline with the test description we should also have

```
expect($interpolate('{{a}}')({ a: date })).not.toEqual('"' + date.toString() + '"');
```

petebacondarwin commented on test/ngMessageFormat/messageFormatSpec.js:
```diff
@@ -329,7 +329,8 @@ describe('$$ngMessageFormat', function() {
     it('should NOT use toString on Date objects', inject(function($interpolate) {
-      expect($interpolate('{{a}}')({ a: new Date(2014, 10, 10, 0, 0, 0) })).toEqual('"2014-11-09T23:00:00.000Z"');
+      var date = new Date(2014, 10, 10);
+      expect($interpolate('{{a}}')({ a: date })).toBe('"' + date.toJSON() + '"');
```
Similarly here?

petebacondarwin commented:
The nitpick was https://github.com/angular/angular.js/pull/14715#discussion_r68054276 but it looks like it is dealt with.

LGTM

petebacondarwin commented:
Now that this is not a breaking change, right? Can we backport it to 1.5 too?

Narretz commented:
I think respecting a custom toString fn could be a breaking change for some.

I've also marked the ngBind change as breaking, but I guess this is more of a fix that now has the same consequences as using toString with interpolate.

gkalpak commented on docs/content/guide/interpolation.ngdoc:
```diff
@@ -26,6 +26,15 @@ normal {@link ng.$rootScope.Scope#$digest digest} cycle.
+If something other than a String is interpolated, the string is computed as follows:
+- If the object is not a `Number`, `Date` or `Array`, $interpolate looks for a custom `toString()`
+function on the object, and uses that. Custom means that
+`myObject.toString() !== `Object.prototype.toString()`.
```
This is not accurate and may confuse people. You mean `myObject.toString !== Object.prototype.toString`.

Narretz replied:
Thanks that's right

gkalpak commented on docs/content/guide/interpolation.ngdoc:
```diff
@@ -26,6 +26,15 @@ normal {@link ng.$rootScope.Scope#$digest digest} cycle.
+function on the object, and uses that. Custom means that
+`myObject.toString() !== `Object.prototype.toString()`.
+
+- If the above doesn't apply, then `JSON.stringify` is used.
```
There is no mention about `null`/`undefined` being converted to the empty string.

gkalpak commented on src/Angular.js:
```diff
@@ -1903,6 +1904,27 @@ function createMap() {
+      value = '' + value;
+      break;
+    default:
+      if (hasCustomToString(value) && !isArray(value) && !isDate(value)) {
```
I am not sure I am a big fan of not `toString`ifying Dates.

Narretz replied:
They were jsonified before and that output is more universal than toString, which produces an American date format while jsonified is ISO.

gkalpak replied:
They were JSONified in `$interpolate`, but `toString`ified in `ngBind`. Since we are making the two consistent, we need to break one of them.

But you convicned me: a universal format is better than forcing American English for all users.
(For some reason, I was under the impression `toString()` returned a localized string...)

Narretz replied:
I thought that too (localized string) and might have written that somewhere.

gkalpak commented on src/ngMessageFormat/messageFormatCommon.js:
```diff
@@ -8,15 +8,7 @@
 /* global isFunction: false */
 /* global noop: false */
 /* global toJson: false */
```
I think this is not necessary any more (but it may have gone away in master anyway).

gkalpak commented on test/ng/directive/ngBindSpec.js:
```diff
@@ -46,6 +46,42 @@ describe('ngBind*', function() {
+      $rootScope.bind = new Date(2014, 10, 10, 0, 0, 0);
+      element = $compile('<div ng-bind="bind"></div>')($rootScope);
+      $rootScope.$digest();
+      expect(element.text()).toBe('"' + $rootScope.bind.toJSON() + '"');
```
Why not `JSON.stringify($rootScope.bind)`?

gkalpak commented on test/ng/interpolateSpec.js:
```diff
@@ -35,6 +35,28 @@ describe('$interpolate', function() {
+
+  it('should NOT use toString on Date objects', inject(function($interpolate) {
+    var date = new Date(2014, 10, 10);
+    expect($interpolate('{{a}}')({ a: date })).toBe('"' + date.toJSON() + '"');
```
Same as above.

gkalpak commented:
Both changes (to `ngBind` and `$interpolate`) are breaking changes imo.
Currently, the `ngBind` is a much bigger change, since many people might be affected by calling `toJson` on Date objects instead of it's `toString` method.

(TBH, I think it is marginally better to call `toString` on Dates (instead of `toJson`), although that would be a significant change for `$interpolate`.)

Narretz commented:
Updated again, should be good to go now

gkalpak commented on docs/content/guide/interpolation.ngdoc:
```diff
@@ -26,6 +26,14 @@ normal {@link ng.$rootScope.Scope#$digest digest} cycle.
+- `undefined` and `null` are converted to `''`
+- if the value is an object that is not a `Number`, `Date` or `Array`, $interpolate looks for
+a custom `toString()` function on the object, and uses that. Custom means that
+`myObject.toString !== `Object.prototype.toString`. If that doesn't apply, `JSON.stringify` is used.
```
> If that doesn't apply, `JSON.stringify` is used.

Nitpick: I would place that on it's own bullet.

gkalpak commented:
LGTM

BTW, in the commit message for the `ngBind` change, the proposed migration instructions (using a `toString` filter) is not 100% correct, as it will not handle `undefined`/`null` values. Besides, I believe a better/easier migration would be using `value.toString()` (which takes advantage of Angular's "forgiving-ness" for `undefined`/`null` expressions.

Narretz commented:
@gkalpak Okay, so ngBind previously set `undefined` to empty string, and `null` was handled by setting textContent, which also sets it to '' (I think). So in the migration note I can special case both of them.

gkalpak commented:
You jave to special case them if you propose using a filter. But I think a filter is not necessary in this case (you can just call `.toString` (that's all the filter does anyway)).

shmdhussain commented:
What is the status of this PR, I believe it is merged into master branch... but I am getting he inconsistent results when interpolating the object, sometime getting full object some time getting `[object Object]`

for example

```
{{ {id:123}  }} //sometime outputs '{id:123}'
{{ {id:123}  }} //sometime outputs [object Object]
```

gkalpak commented:
This is merged and included in all version since 1.6.0-rc.0 onwards.
Even before this PR, it shouldn't cause inconsistent behavior. Your problem is caused by something else.

shmdhussain commented:
See this screenshot, sometimes I am getting the stringified version of Object sometime getting `[object Object]`
![Screenshot 2020-02-24 at 17 59 50](https://user-images.githubusercontent.com/4039657/75158485-a9149800-572f-11ea-8baf-f13609940cbd.png)
